packages/flutter/test/widgets/implicit_animations_test.dart

```diff
@@ -47,4 +76,154 @@ void main() {
     final Matrix4 result = tween.lerp(0.25);
     expect(result, equals(Matrix4.translationValues(11.0, 21.0, 31.0)));
   });
+
+  testWidgets('onEnd callback test', (WidgetTester tester) async {
+    int i = 0;
+    for (ImplicitAnimatedWidgetType widgetType
+        in ImplicitAnimatedWidgetType.values) {
+      await tester.pumpWidget(wrap(
+          child: TestAnimatedWidget(mockOnEndFunction, switchKey, widgetType)));
+
+      final Finder widgetFinder = find.byKey(switchKey);
+
+      await tester.tap(widgetFinder);
+
+      await tester.pumpAndSettle();
```

Comment:
I would use a for loop to pump frame one by one and make sure onEnd only called at the very end
```dart
// Assume animation duration is 1 second.
await tester.pump();
await tester.pump(duration: Duration(milliseconds:200));
expect(mockOnEndFunction.called, i);
await tester.pump(duration: Duration(milliseconds:200));
expect(mockOnEndFunction.called, i);
await tester.pump(duration: Duration(milliseconds:200));
expect(mockOnEndFunction.called, i);
await tester.pump(duration: Duration(milliseconds:200));
expect(mockOnEndFunction.called, i);
await tester.pump(duration: Duration(milliseconds:200));
expect(mockOnEndFunction.called, ++i);
```

Reply:
I guess this should be sufficient:  
```
    await tester.pump();
    expect(mockOnEndFunction.called, 0);
    await tester.pump(const Duration(milliseconds:1000));
    expect(mockOnEndFunction.called, 0);
    await tester.pump(const Duration(milliseconds:1));
    expect(mockOnEndFunction.called, 1);
```